common/app/views/support/cleaner/AmpAdCleaner.scala

```diff
@@ -12,84 +12,80 @@ import scala.xml.Elem
 
 object AmpAdCleaner {
   val AD_LIMIT = 8
-  val CHARS_BETWEEN_ADS = 700
-  val DONT_INTERLEAVE_SMALL_PARA = 50
-
-  def findElementsNeedingAdsAfter(children: List[Element]): List[Element] = {
-
-    val ALLOWED = 0
-    val DISALLOWED = 1
-
-    // create the constraints for each paragraph how far before and after we can have it
-    val constraints = children.map { element =>
-
-      def para = element.tagName() != "p"
-      def short = element.text().length < DONT_INTERLEAVE_SMALL_PARA
-
-      if (para) {
-        // don't put an ad before or too close after an embed
-        AdRepel(300, 200, element)
-      } else if (short) {
-        // don't interleave ads between small paragraphs
-        AdRepel(DISALLOWED, DISALLOWED, element)
-      } else {
-        AdRepel(ALLOWED, ALLOWED, element)
+  val SMALL_PARA_CHARS = 50
+  val MIN_CHAR_BUFFER = 700
+  val IMG_BUFFER_FWD = 300 // really any non-p element type
+  val IMG_BUFFER_BWD = 200
+
+  /**
+    * Find ad slots
+    *
+    * Returns a list of elements *after* which ads should be directly
+    * placed.
+    *
+    * Ads are placed:
+    *
+    * * sufficiently far from other ads (MIN_CHAR_BUFFER characters away)
+    * * sufficiently far from non-text (p) elements (IMG_BUFFER_[FWD|BWD])
+    * * non adjacent to small (SMALL_PARA_CHARS) paragraphs
+    *
+    * These tests apply forwards and backwards, though in the case of
+    * non-text buffers, the values differ.
+    *
+    * Where the above tests are met, we say there is adequate 'buffer'
+    * around the ad and it can be placed.
+    */
+  def findAdSlots(elems: Vector[Element]): List[Element] = {
+    def isPara(elem: Element): Boolean = elem.tagName() == "p"
+    def suitableAdNeighbour(elem: Element): Boolean = isPara(elem) && elem.text.length > SMALL_PARA_CHARS
+
+    def hasForwardBuffer(elems: Vector[Element], index: Int): Boolean = {
+      val enoughCharsFwd = {
+        val fwdElems = elems.drop(index + 1)
+        val meetsThreshold = fwdElems.takeWhile(_.tagName == "p").map(_.text.length).sum >= IMG_BUFFER_FWD
+        meetsThreshold || !fwdElems.exists(_.tagName != "p")
```

Comment:
the second part of this could be `fwdElems.forall(_.tagName == "p")`

maybe it could be stored in a val called `noForwardEmbeds` or something?